ts-client/src/farm.ts

```diff
@@ -0,0 +1,408 @@
+import { BN, EventParser } from "@coral-xyz/anchor";
+import { TOKEN_PROGRAM_ID } from "@solana/spl-token";
+import {
+  Cluster,
+  ComputeBudgetProgram,
+  Connection,
+  PublicKey,
+  Transaction,
+  TransactionInstruction,
+} from "@solana/web3.js";
+
+import { FarmProgram, Opt, PoolState, UserState } from "./types";
+import {
+  chunks,
+  getFarmInfo,
+  getFarmProgram,
+  getOrCreateATAInstruction,
+  parseLogs,
+} from "./utils";
+import { FARM_PROGRAM_ID, SIMULATION_USER } from "./constant";
+
+const chunkedFetchMultipleUserAccount = async (
+  program: FarmProgram,
+  pks: PublicKey[],
+  chunkSize: number = 100
+) => {
+  const accounts = (
+    await Promise.all(
+      chunks(pks, chunkSize).map((chunk) =>
+        program.account.user.fetchMultiple(chunk)
+      )
+    )
+  ).flat();
+
+  return accounts.filter(Boolean);
+};
+
+const chunkedFetchMultiplePoolAccount = async (
+  program: FarmProgram,
+  pks: PublicKey[],
+  chunkSize: number = 100
+) => {
+  const accounts = (
+    await Promise.all(
+      chunks(pks, chunkSize).map((chunk) =>
+        program.account.pool.fetchMultiple(chunk)
+      )
+    )
+  ).flat();
+
+  return accounts.filter(Boolean);
+};
+
+const getAllPoolState = async (
+  farmMints: Array<PublicKey>,
+  program: FarmProgram
+) => {
+  const poolStates = (await chunkedFetchMultiplePoolAccount(
+    program,
+    farmMints
+  )) as Array<PoolState>;
+
+  return poolStates;
+};
+
+const MAX_CLAIM_ALL_ALLOWED = 2;
+
+export class PoolFarmImpl {
+  private opt: Opt = {
+    cluster: "mainnet-beta",
+  };
+
+  private constructor(
+    public address: PublicKey,
+    private program: FarmProgram,
+    private eventParser: EventParser,
+    public poolState: PoolState,
+    opt: Opt
+  ) {
+    this.opt = opt;
+  }
+
+  public static async createMultiple(
+    connection: Connection,
+    farmList: Array<PublicKey>,
+    opt?: {
+      cluster?: Cluster;
+    }
+  ) {
+    const cluster = opt?.cluster ?? "mainnet-beta";
+    const { program } = getFarmProgram(connection);
+    const eventParser = new EventParser(FARM_PROGRAM_ID, program.coder);
+
+    const poolsState = await getAllPoolState(farmList, program);
+
+    return poolsState.map((poolState, idx) => {
+      const address = farmList[idx];
+      return new PoolFarmImpl(address, program, eventParser, poolState, {
+        cluster,
+      });
+    });
+  }
+
+  public static async getFarmAddressByPoolAddress(
+    poolAddress: PublicKey,
+    cluster?: Cluster
+  ) {
+    const apiData = await getFarmInfo(cluster);
+
+    const farm = apiData.find(
+      (farm) => farm.pool_address === poolAddress.toBase58()
+    );
+
+    if (!farm) throw new Error("No pool address found ");
+
+    return {
```

Comment:
We shoudl return array here, because there may have more farming pools for a poolAddress